src/runtime/crdt/crdt-collection.ts

```diff
@@ -51,32 +60,85 @@ type CollectionModel<T extends Referenceable> = CRDTModel<CRDTCollectionTypeReco
 export class CRDTCollection<T extends Referenceable> implements CollectionModel<T> {
   private model: CollectionData<T> = {values: {}, version: {}};
 
-  merge(other: CollectionData<T>):
-      {modelChange: CollectionChange<T>, otherChange: CollectionChange<T>} {
-    const newValues = this.mergeItems(this.model, other);
-    const newVersion = mergeVersions(this.model.version, other.version);
-    this.model.values = newValues;
-    this.model.version = newVersion;
-    // For now this is always returning a model change.
-    const change: CollectionChange<T> = {
+  merge(other: CollectionData<T>): {modelChange: CollectionChange<T>, otherChange: CollectionChange<T>} {
+    const newClock = mergeVersions(this.model.version, other.version);
+    const merged: Dictionary<{value: T, version: VersionMap}> = {};
+
+    // Fast-forward op to send to other model. Elements added and removed will
+    // be filled in below.
+    const fastForwardOp: CollectionFastForwardOp<T> = {
+      type: CollectionOpTypes.FastForward,
+      added: [],
+      removed: [],
+      oldClock: other.version,
+      newClock,
+    };
+
+    for (const otherEntry of Object.values(other.values)) {
+      const value = otherEntry.value;
+      const id = value.id;
+      const thisEntry = this.model.values[id];
+
+      if (thisEntry) {
+        if (sameVersions(thisEntry.version, otherEntry.version)) {
+          // Both models have the same value at the same version. Add it to the
+          // merge.
+          merged[id] = thisEntry;
+        } else {
+          // Models have different versions for the same value. Merge the
+          // versions, and update other.
+          const mergedVersion = mergeVersions(thisEntry.version, otherEntry.version);
+          merged[id] = {value, version: mergedVersion};
+          fastForwardOp.added.push([value, mergedVersion]);
+        }
+      } else if (dominates(this.model.version, otherEntry.version)) {
+        // Value was deleted by this model.
+        fastForwardOp.removed.push(value);
+      } else {
+        // Value was added by other model.
+        merged[id] = otherEntry;
+      }
+    }
+    for (const thisEntry of Object.values(this.model.values)) {
+      const id = thisEntry.value.id;
```

Comment:
can this iterate over Object.entries?
for (const [thisEntry, id] of Object.entries(this.model.values)) {